### PR TITLE
upload for review

### PR DESCRIPTION
**Pull request by RRoundTable**
Thanks!

### CODE REVIEW

File: review/Copy List_with_Random_Pointer.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        self.random = random
+
+
+class Solution:
+    def copyRandomList(self, head: 'Node') -> 'Node':
+        '''O(N)/O(N)
+        Iterative
+        '''
+        d, ans, start = {None: [None, None, None]}, None, head
+        while head != None:
```
Comment (Curt-Park) [inline review/Copy List_with_Random_Pointer.py]: Just use `while head`

File: review/Happy_Number.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    def isHappy(self, n):
+        """
+        :type n: int
+        :rtype: bool
+        """
+        d = {}
+        while d.get(n, True):
+            if n == 1: return True
+            d[n] = False
+            n = sum([int(s) ** 2 for s in str(n)])
```
Comment (Curt-Park) [inline review/Happy_Number.py]: you can use `n = sum(int(s) ** 2 for s in str(n))` either.

File: review/Happy_Number.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    
+    def isHappy(self, n):
+        """
+        :type n: int
+        :rtype: bool
+        """
+        if self.seen[n]: return False
+        self.seen[n], res = True, 0
+        while n != 0: 
+            res += (n % 10) ** 2; n = int(n / 10)
```
Comment (Curt-Park) [inline review/Happy_Number.py]: See `divmod` builtin method.

File: review/Happy_Number.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        return True if res == 1 else self.isHappy(res)
```
Comment (Curt-Park) [inline review/Happy_Number.py]: `return res == 1 or self.isHappy(res)` is also fine.

File: review/Insert_ Delete_GetRandom_O(1).py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+    def remove(self, val: int) -> bool:
+        """
+        Removes a value from the set. Returns true if the set contained the specified element.
+        """
+        if self.dict.get(val, -1) != -1:
+            idx, last_element = self.dict[val], self.list[-1]
+            self.list[idx], self.dict[last_element] = last_element, idx
+            self.list.pop()
+            del self.dict[val]
```
Comment (Curt-Park) [inline review/Insert_ Delete_GetRandom_O(1).py]: just for your information:
https://stackoverflow.com/questions/5713218/best-method-to-delete-an-item-from-a-dict

File: review/Insert_ Delete_GetRandom_O(1).py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+import random
+
+
+class RandomizedSet:
+
+    def __init__(self):
+        """
+        Initialize your data structure here.
+        """
+        self.dict = {}
```
Comment (Curt-Park) [inline review/Insert_ Delete_GetRandom_O(1).py]: just for your information.
we can type these lines as follows:
```
self.dict: Dict[int, int] = {}
self.list: List[int] = []
```

File: review/Insert_ Delete_GetRandom_O(1).py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        self.list = []
+
+    def insert(self, val: int) -> bool:
+        """
+        Inserts a value to the set. Returns true if the set did not already contain the specified element.
+        """
+        if self.dict.get(val, -1) == -1:
```
Comment (Curt-Park) [inline review/Insert_ Delete_GetRandom_O(1).py]: `if self.dict.get(val, False):` looks better.

File: review/Longest_Substring_Without_Repeating_Characters.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+Output: 3
+Explanation: The answer is "wke", with the length of 3. 
+             Note that the answer must be a substring, "pwke" is a subsequence and not a substring.
+'''
+
+class Solution:
+    def lengthOfLongestSubstring(self, s: str) -> int:
+        d, ans, i, j = {}, 0, -1, 0
+        while j < len(s):
+            if d.get(s[j], -1) == -1:
```
Comment (Curt-Park) [inline review/Longest_Substring_Without_Repeating_Characters.py]: `d.get(s[j], False)`

File: review/Longest_Substring_Without_Repeating_Characters.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+                d[s[j]], j = j, j + 1
```
Comment (Curt-Park) [inline review/Longest_Substring_Without_Repeating_Characters.py]: if you use `continue` in the next line, you can remove `else` and the indentation.

File: review/Minimum_Window_Substring.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        ans, count, l, r, char, s = [float('inf'), ''], len(t) , 0, 0, Counter(t), s + ' '
+        while r < len(s):
+            if count > 0:
+                if char.get(s[r], -float('inf')) > -float('inf'):
+                    if char.get(s[r], -float('inf')) > 0: count -= 1
+                    char[s[r]] -= 1
+                r += 1
+            else:
+                if ans[0] > r - l:
+                    ans = [r - l, s[l:r]]
```
Comment (Curt-Park) [inline review/Minimum_Window_Substring.py]: slicing makes a copy of the sublist.
How about storing the indices?

File: review/Minimum_Window_Substring.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+                if char.get(s[l], -float('inf')) > -float('inf'):
```
Comment (Curt-Park) [inline review/Minimum_Window_Substring.py]: float('-inf') looks better.

Comment (Curt-Park) [reply review/Minimum_Window_Substring.py]: or just
`if char.get(s[l], False):` or
`if s[i] in char:`

File: review/Minimum_Window_Substring.py [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+                    if char.get(s[l], -float('inf')) >= 0: count += 1
```
Comment (Curt-Park) [inline review/Minimum_Window_Substring.py]: another expression:
`count += s[l] in char`

File: review/Copy List_with_Random_Pointer.py (hunk 1 above)
Comment (RRoundTable) [reply review/Copy List_with_Random_Pointer.py]: Thanks!

File: review/Happy_Number.py (hunk 2 above)
Comment (RRoundTable) [reply review/Happy_Number.py]: I didn't know `int(s) ** 2 for s in str(n)` is generator.!
Thank you!

File: review/Happy_Number.py (hunk 4 above)
Comment (RRoundTable) [reply review/Happy_Number.py]: It looks more simple!

File: review/Happy_Number.py (hunk 3 above)
Comment (RRoundTable) [reply review/Happy_Number.py]: Good information!

File: review/Insert_ Delete_GetRandom_O(1).py (hunk 6 above)
Comment (RRoundTable) [reply review/Insert_ Delete_GetRandom_O(1).py]: It looks easy to understand.

File: review/Insert_ Delete_GetRandom_O(1).py (hunk 5 above)
Comment (RRoundTable) [reply review/Insert_ Delete_GetRandom_O(1).py]: Thanks to you, I can use pop in dict!

File: review/Insert_ Delete_GetRandom_O(1).py (hunk 7 above)
Comment (RRoundTable) [reply review/Insert_ Delete_GetRandom_O(1).py]: I agree. But in that case, I think 0-th index is considered `False` even if it is not False.

File: review/Longest_Substring_Without_Repeating_Characters.py (hunk 9 above)
Comment (RRoundTable) [reply review/Longest_Substring_Without_Repeating_Characters.py]: Thanks!

File: review/Longest_Substring_Without_Repeating_Characters.py (hunk 8 above)
Comment (RRoundTable) [reply review/Longest_Substring_Without_Repeating_Characters.py]: I think 0-th index is considered False even if it is not False.

File: review/Minimum_Window_Substring.py (hunk 10 above)
Comment (RRoundTable) [reply review/Minimum_Window_Substring.py]: Good point!

File: review/Minimum_Window_Substring.py (hunk 11 above)
Comment (RRoundTable) [reply review/Minimum_Window_Substring.py]: It looks better!
`class Solution:
    '''O(|S|+|T|)/O(|T|)'''
    def minWindow(self, s: str, t: str) -> str:
        ans, count, l, r, char, s = [float('inf'), ''], len(t) , 0, 0, Counter(t), s + ' '
        while r < len(s):
            if count > 0:
                if s[r] in char:
                    if char[s[r]] > 0: count -= 1
                    char[s[r]] -= 1
                r += 1
            else:
                if ans[0] > r - l:
                    ans = [r - l, (l, r)]
                if s[l] in char:
                    if char[s[l]] >= 0: count += 1
                    char[s[l]] += 1
                l += 1
        return s[ans[1][0]:ans[1][1]] if ans[0] < float('inf') else ''`

File: review/Minimum_Window_Substring.py (hunk 12 above)
Comment (RRoundTable) [reply review/Minimum_Window_Substring.py]: Interesting!
```python
def minWindow(self, s: str, t: str) -> str:
        ans, count, l, r, char, s = [float('inf'), ''], len(t) , 0, 0, Counter(t), s + ' '
        while r < len(s):
            if count > 0:
                if s[r] in char:
                    count -= char[s[r]] > 0
                    char[s[r]] -= 1
                r += 1
            else:
                if ans[0] > r - l:
                    ans = [r - l, (l, r)]
                if s[l] in char:
                    count += char[s[l]] >= 0
                    char[s[l]] += 1
                l += 1
        return s[ans[1][0]:ans[1][1]] if ans[0] < float('inf') else ''
```

File: review/Insert_ Delete_GetRandom_O(1).py [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+
+class RandomizedSet:
+
+    def __init__(self):
+        """
+        Initialize your data structure here.
+        """
+        self.dict = {}
+        self.list = []
```
Comment (Curt-Park) [inline review/Insert_ Delete_GetRandom_O(1).py]: `dict` and `list` are keywords. It shouldn't be used as variable names.

**Pull request merged**